extensions/transferdetails/service.go

```diff
@@ -0,0 +1,350 @@
+package transferdetails
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/centrifuge/go-centrifuge/jobs"
+
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/extensions"
+	"github.com/centrifuge/go-centrifuge/identity"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	logging "github.com/ipfs/go-log"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+
+	// UpdateTransferDetail updates a TransferDetail
+	UpdateTransferDetail(ctx context.Context, req UpdateTransferDetailRequest) (documents.Model, jobs.JobID, error)
+
+	// CreateTransferDetail derives a TransferDetail from a request payload
+	CreateTransferDetail(ctx context.Context, req CreateTransferDetailRequest) (documents.Model, jobs.JobID, error)
+
+	// DeriveFundingResponse returns a TransferDetail in client format
+	DeriveTransferDetail(ctx context.Context, model documents.Model, transferID []byte) (*TransferDetail, documents.Model, error)
+
+	// DeriveFundingListResponse returns a TransferDetail list in client format
+	DeriveTransferList(ctx context.Context, model documents.Model) (*TransferDetailList, documents.Model, error)
+}
+
+// service implements Service and handles all funding related persistence and validations
+type service struct {
+	srv           documents.Service
+	tokenRegistry documents.TokenRegistry
+}
+
+const (
+	transfersLabel    = "transfer_details"
+	transfersFieldKey = "transfer_details[{IDX}]."
+	transferIDLabel   = "transfer_id"
+)
+
+// DefaultService returns the default implementation of the service.
+func DefaultService(
+	srv documents.Service,
+	tokenRegistry documents.TokenRegistry,
+) Service {
+	return service{
+		srv:           srv,
+		tokenRegistry: tokenRegistry,
+	}
+}
+
+var log = logging.Logger("transferdetail-api")
+
+// TODO: get rid of this or make generic
+func deriveDIDs(data *TransferDetailData) ([]identity.DID, error) {
+	var c []identity.DID
+	for _, id := range []string{data.SenderID, data.RecipientID} {
+		if id != "" {
+			did, err := identity.NewDIDFromString(id)
+			if err != nil {
+				return nil, err
+			}
+			c = append(c, did)
+		}
+	}
+
+	return c, nil
+}
+
+func (s service) updateModel(ctx context.Context, model documents.Model) (documents.Model, jobs.JobID, error) {
+	cs, err := model.GetCollaborators()
+	if err != nil {
+		return nil, jobs.NilJobID(), err
+	}
+
+	a := model.GetAttributes()
+	attr, err := extensions.ToClientAttributes(a)
+	if err != nil {
+		return nil, jobs.NilJobID(), err
+	}
+
+	d := model.GetData().([]byte)
+
+	payload := documents.UpdatePayload{
+		DocumentID: model.ID(),
+		CreatePayload: documents.CreatePayload{
+			Scheme:        model.Scheme(),
+			Collaborators: cs,
+			Attributes:    attr,
+			Data:          d,
+		},
+	}
+
+	updated, jobID, err := s.srv.UpdateModel(ctx, payload)
```

Comment:
Why we can't use here the `coreapi.Service` instead of the `documents.Service`? From my point of view any userapi logic, should always interact with the `coreapi.Service` and not directly to the underlaying documents service.
Does it trigger a cyclic dep?

Reply:
yes, it does :/

Reply:
ok, this would involve a nice refactoring... lets add a comment there and move on with this implementation.

Reply:
"nice refactoring" ^-^